merkletree.go

```diff
@@ -175,13 +194,44 @@ insertLoop:
 			panic(ErrInvalidTree)
 		}
 	}
-	return
+}
+
+func (m *MerkleTree) recomputeHash() {
+	if m.root.leftHash == nil {
+		m.root.leftHash = m.hashNode(m.root.leftChild)
+	}
+	if m.root.rightHash == nil {
+		m.root.rightHash = m.hashNode(m.root.rightChild)
+	}
+	m.treeHash = m.root.hash()
+}
+
+func (m *MerkleTree) GetHash() []byte {
+	return m.treeHash
+}
+
+func (m *MerkleTree) hashNode(node MerkleNode) []byte {
+	switch node.(type) {
+	case *interiorNode:
+		currentNodeI := node.(*interiorNode)
+		if currentNodeI.leftHash == nil {
+			currentNodeI.leftHash = m.hashNode(currentNodeI.leftChild)
+		}
+		if currentNodeI.rightHash == nil {
+			currentNodeI.rightHash = m.hashNode(currentNodeI.rightChild)
+		}
```

Comment:
Is there any reason not to move,

```
+		if currentNodeI.leftHash == nil {
 +			currentNodeI.leftHash = m.hashNode(currentNodeI.leftChild)
 +		}
 +		if currentNodeI.rightHash == nil {
 +			currentNodeI.rightHash = m.hashNode(currentNodeI.rightChild)
 +		}
```

into `func (n *interiorNode) hash() []byte {`?

In which case, this all simplifies down to `node.hash(m)` and the `hashNode()` function is not needed.